tracing/test-agent-shutdown.sh

```diff
@@ -0,0 +1,1341 @@
+		'local') true ;;
+
+		# Extract journal entries for the duration of the test
+		'vm')
+			sudo journalctl \
+				-q \
+				-a \
+				-o cat \
+				-t 'kata' \
+				--since="$test_start_time" \
+				> "$log_file"
+		;;
+
+		*) die "invalid agent test type: '$agent_test_type'" ;;
+	esac
+
+	[ -e "$log_file" ] || die "no log file: '$log_file'"
+	[ -s "$log_file" ] || die "empty log file: '$log_file'"
+
+	true
+}
+
+# Function to run to ensure correct behaviour
+validate_agent()
+{
+	local agent_test_type="${1:-}"
+	local shutdown_test_type="${2:-}"
+	local log_file="${3:-}"
+
+	[ -z "$agent_test_type" ] && die "need agent test type"
+	[ -z "$shutdown_test_type" ] && die "need shutdown test type"
+	[ -z "$log_file" ] && die "need agent log file"
+
+	info "validating"
+
+	get_agent_log_file \
+		"$agent_test_type" \
+		"$log_file"
+
+	# Regular expression that describes possible agent failures
+	local regex="(slog::Fuse|Drain|Custom|serialization error|thread.*panicked|stack backtrace:)"
+
+	egrep -qi "$regex" "$log_file" && die "Found agent error in log file: '$log_file'"
+
+	local entry
+	entry=$(get_shutdown_test_type_entry "$shutdown_test_type" || true)
+	[ -z "$entry" ] && die "invalid test type: '$shutdown_test_type'"
+
+	local hypervisor_debug=$(echo "$entry"|cut -d: -f3)
+	local vsock_console=$(echo "$entry"|cut -d: -f4)
+
+	local agent_debug_logs_available='false'
+
+	[ "$hypervisor_debug" = 'true' ] && \
+		[ "$vsock_console" = 'false' ] && \
+		agent_debug_logs_available='true'
+
+	if [ "$agent_debug_logs_available" = 'true' ] || [ "$agent_test_type" = 'local' ]
+	then
+		# The message the agent writes to stderr just before it exits.
+		local done_msg="\<shutdown complete\>"
+
+		egrep -q "$done_msg" "$log_file" || die "missing agent shutdown message"
+	else
+		# We can only check for the shutdown message if the agent debug
+		# logs are available.
+		info "Not checking for agent shutdown message as hypervisor debug disabled"
+	fi
+}
+
+setup_agent()
+{
+	local shutdown_test_type="${1:-}"
+	[ -z "$shutdown_test_type" ] && die "need shutdown test type"
+
+	configure_kata "$shutdown_test_type"
+
+	true
+}
+
+check_agent_stopped()
+{
+	info "Checking agent stopped"
+
+	local agent_test_type="${1:-}"
+	[ -z "$agent_test_type" ] && die "need agent test type"
+
+	local cmd=
+
+	case "$agent_test_type" in
+		'local') cmd=check_local_agent_stopped ;;
+		'vm') cmd=check_vm_stopped ;;
+		*) die "invalid agent test type: '$agent_test_type'" ;;
+	esac
+
+	waitForProcess \
+		"$shutdown_time_secs" \
+		"$sleep_time_secs" \
+		"$cmd"
+
+	true
+}
+
+check_local_agent_stopped()
+{
+	local ret=0
+
+	local i=0
+	local max=20
+
+	agent_ended="false"
+
+	local agent_pid
+	agent_pid=$(get_local_agent_pid || true)
+
+	# Agent has finished
+	[ -z "$agent_pid" ] && return 0
+
+	for i in $(seq "$max")
+	do
+		{ sudo kill -0 "$agent_pid"; ret=$?; } || true
+
+		[ "$ret" -ne 0 ] && agent_ended="true" && break
+
+		sleep 0.2
+
+		i=$((i+1))
```

Comment:
is this really needed ? I don't think so.. :thinking:

Reply:
Good catch - fixed.